File: Practica1_i-Lunch/src/gestorAplicacion/usuariosRestaurante/Mesero.java
/*CLASE CREADA POR JERONIMO GOMEZ RESTREPO
  Clase que hereda de EMPLEADO y que tiene como finalidad tomar los pedidos(crear?), llevar un pedido a su etapa final "Entregado"
  y dar a conocer el precio total del pedido.
  Por ahora no se agregan nuevos atributos. 
  */

package gestorAplicacion.usuariosRestaurante;

import java.util.ArrayList;

import gestorAplicacion.gestionRestaurante.*;
import uiMain.DatosAleatorios;

public class Mesero extends Empleado implements Usuario {

	// Serialización
	private static final long serialVersionUID = 1L;
	private static ArrayList<Mesero> meseros;
	static {
		meseros = new ArrayList<Mesero>();
	}

	// Aca se almacenan todos los pedidos atentidos por el mesero en cuestion

	private ArrayList<Pedido> pedidosAtendidos = new ArrayList<Pedido>();
	private ArrayList<Integer> historialPropinas = new ArrayList<Integer>();

	/* Constructor de la clase Mesero */
	public Mesero(int cedula, String nombre, boolean disponibilidad, int salario,
			Restaurante restaurante/* , boolean pagado */) {
		this.setCedula(cedula);
		this.setNombre(nombre);
		this.setCargo("Mesero");
		this.setDisponibilidad(disponibilidad);
		this.setSalario(salario);
		this.setRestaurante(restaurante);
		/* this.setPagado = pagado; */

		// Serializacion
		ArrayList<Empleado> empleados = Empleado.getEmpleados();
		empleados.add(this);
		meseros.add(this);
		Empleado.setEmpleados(empleados);
	}

	/* Sobrecarga del constructor para valores predeterminados */
	public Mesero() {
		this(0, "NN", false, 0, null/* , false */);
	}

	// Getters y Setters
	public static ArrayList<Mesero> getMeseros() {
		return meseros;
	}

	/*
	 * Metodo para cambiar el estado de un pedido(atributo "estadoPedido") a su
	 * estado final "Entregado". Recibe como parametro un pedido y tiene como
	 * finalidad modificar el atributo "estadoPedido", por ultima vez y de manera
	 * unica. Se debe revisar que el estado anterior del pedido sea
	 * "Listo para ser despachado" y que sea para consumir en el lugar.
	 */

	public void llevarPedido(Pedido pedido) {
		pedidosAtendidos.add(pedido); // Se agrega el pedido a la lista de pedidos
		if (pedido.getEstado().equals("Listo")
				&& pedido.getTipo().equals("Para Consumir en el lugar")) { /* <<<<<--------- Falta aplicar el ENUM */
			pedido.setEstado("Despachado"); /* <<<<<--------- Falta aplicar el ENUM */
		}
	}

	// metodo para agregar una propina al historial de propinas

	public void recibirPropina(int propina) {
		historialPropinas.add(propina);
	}

	// Metodo que calcula el total de las propinas recibidas por un mesero.
	public int totalPropinas() {
		int propinaAdder = 0;
		for (float propina : historialPropinas) {
			propinaAdder += propina;
		}
		return propinaAdder;
	}

	//// GETTERS AND SETTERS \\\\
	public ArrayList<Pedido> getPedidosAtendidos() {
		return pedidosAtendidos;
	}

	public void setPedidosAtendidos(ArrayList<Pedido> pedidosAtendidos) {
		this.pedidosAtendidos = pedidosAtendidos;
	}

	public static void setMeseros(ArrayList<Mesero> meseros) {
		Mesero.meseros = meseros;
	}

	public ArrayList<Integer> getHistorialPropinas() {
		return historialPropinas;
	}

	public void setHistorialPropinas(ArrayList<Integer> historialPropinas) {
		this.historialPropinas = historialPropinas;
	}

	// Implementación de la interfaz Usuario
	public String informacion() {
		if (this.getDisponibilidad()) {
			return "El Mesero " + this.nombre + " con C.C. " + this.cedula + " trabaja en el restaurante "
					+ this.restaurante.getNombre() + "\n" + "Tiene un salario de: $" + this.salario + "\n"
					+ "Está disponible actualmente.";
		} else {
			return "El Mesero " + this.nombre + " con C.C. " + this.cedula + " trabaja en el restaurante "
					+ this.restaurante.getNombre() + "\n" + "Tiene un salario de: $" + this.salario + "\n"
					+ "No está disponible actualmente.";
		}
	}

	public void agregarPedidoHistorial(Pedido pedido) {
		pedidosAtendidos.add(pedido);
		historialPropinas.add(DatosAleatorios.randInt(1, 10));
	}
}

File: Practica2_i-Lunch/src/gestorAplicacion/usuariosRestaurante/mesero.py
from gestorAplicacion.usuariosRestaurante.empleado import Empleado

from gestorAplicacion.gestionRestaurante.estadoPedido import EstadoPedido
from gestorAplicacion.gestionRestaurante.tipoPedido import TipoPedido

class Mesero(Empleado):

    # Atributos estaticos

    _meseros = []

    # Constructor

    def __init__(self, cedula = 0, nombre = "", disponibilidad = False, salario = 0, restaurante = None):
        super().__init__(cedula, nombre, "Mesero", disponibilidad, salario, restaurante)

        self._pedidosAtendidos = []
        self._historialPropinas = []

        Empleado._empleados.append(self)
        Mesero._meseros.append(self)
    
    # Getters y Setters

    def getPedidosAtendidos(self):
        return self._pedidosAtendidos

    def setPedidosAtendidos(self, pedidosAtendidos):
        self._pedidosAtendidos = pedidosAtendidos

    def getHistorialPropinas(self):
        return self._historialPropinas

    def setHistorialPropinas(self, historialPropinas):
        self._historialPropinas = historialPropinas

    @classmethod
    def getMeseros(cls):
        return cls._meseros

    @classmethod
    def setMeseros(cls, meseros):
        cls._meseros = meseros

    # Metodos

    def recibirPropina(self, propina):
        self._historialPropinas.append(propina)

    def agregarPedidoHistorial(self, pedido, propina):
        self._pedidosAtendidos.append(pedido)
        self.recibirPropina(propina)

    # Simulacion de llevar el pedido a un cliente
    def llevarPedido(self, pedido, propina):
        self.agregarPedidoHistorial(pedido, propina)
        
        # Solo se puede hacer si el pedido es de tipo consumir en el local y ya esta listo. Cambiar el estado del pedido
        if pedido.getEstado() == EstadoPedido.LISTO.value and pedido.getTipo() == TipoPedido.TIENDA.value:
            pedido.setEstado(EstadoPedido.DESPACHADO.value)

    # Calcular propinas totales del mesero
    def totalPropinas(self):
        total = 0

        for propina in self._historialPropinas:
            total += propina

        return total

    # Implementacion de la interfaz

    def informacion(self):
        info = f"El Mesero {self._nombre} con C.C. {self._cedula} trabaja en el restaurante {self._restaurante.getNombre()}.\n" \
               f"Tiene un salario de: ${self._salario}.\n"
        
        if self._disponibilidad:
            return info + "Esta disponible actualmente."
        else:
            return info + "No esta disponible actualmente."